dali/pipeline/operators/reader/loader/lmdb.h

```diff
@@ -24,71 +24,154 @@
 
 namespace dali {
 
-#define CHECK_LMDB(status, filename) \
+#define CHECK_LMDB(status) \
   do { \
     DALI_ENFORCE(status == MDB_SUCCESS, "LMDB Error: " + string(mdb_strerror(status)) + \
-                                        ", with file: " + filename); \
+                                        ", with file: " + db_path_); \
   } while (0)
 
-namespace lmdb {
-  inline bool SeekLMDB(MDB_cursor* cursor, MDB_cursor_op op, MDB_val* key, MDB_val *value,
-                       const string &filename) {
-    int status = mdb_cursor_get(cursor, key, value, op);
 
-    if (status == MDB_NOTFOUND) {
-      // reached the end of the db
-      return false;
+class IndexedLMDB {
+  MDB_env* mdb_env_ = nullptr;
+  MDB_cursor* mdb_cursor_ = nullptr;
+  MDB_dbi mdb_dbi_;
+  MDB_txn* mdb_transaction_ = nullptr;
+  int num_;
+  Index mdb_index_;
+  std::string db_path_;
+  Index mdb_size_;
+
+ public:
+  bool Open(const std::string& path, int num) {
+    db_path_ = path;
+    num_ = num;
+    CHECK_LMDB(mdb_env_create(&mdb_env_));
+    auto mdb_flags = MDB_RDONLY | MDB_NOTLS | MDB_NOLOCK;
+    CHECK_LMDB(mdb_env_open(mdb_env_, path.c_str(), mdb_flags, 0664));
+
+    // Create transaction and cursor
+    CHECK_LMDB(mdb_txn_begin(mdb_env_, NULL, MDB_RDONLY, &mdb_transaction_));
+    CHECK_LMDB(mdb_dbi_open(mdb_transaction_, NULL, 0, &mdb_dbi_));
+    CHECK_LMDB(mdb_cursor_open(mdb_transaction_, mdb_dbi_, &mdb_cursor_));
+    MDB_stat stat;
+    CHECK_LMDB(mdb_stat(mdb_transaction_, mdb_dbi_, &stat));
+    mdb_size_ = stat.ms_entries;
+    printf("lmdb file %d %s has %ld entries\n", num_, path.c_str(), mdb_size_);
+    mdb_index_ = 0;
+    return true;
+  }
+  size_t GetSize() const { return mdb_size_; }
+  Index GetIndex() const { return mdb_index_; }
+  bool SeekByIndex(Index index, MDB_val* key = 0, MDB_val* value = 0) {
```

Comment:
You can have `void`/not return anything here, since you don't check the return value.

Reply:
Thank you for your review.